algorithms/kernel/k_nearest_neighbors/kdtree_knn_impl.i

```diff
@@ -80,93 +80,92 @@ inline const T & max(const T & a, const T & b)
     Stack() : _data(nullptr) {}
 
-    ~Stack()
+    void clear()
     {
-        services::daal_free(_data);
-        _data = nullptr;
+        if (_data != nullptr)
+        {
+            services::daal_free(_data);
+            _data = nullptr;
+        }
     }
 
-    bool init(size_t size)
+    ~Stack()
     {
-        _data = static_cast<T *>(services::internal::service_malloc<T, cpu>(size * sizeof(T)));
-        _size = size;
-        _top = _sizeMinus1 = size - 1;
-        _count             = 0;
-        return _data;
+        clear();
     }
 
-    void clear()
+    bool init(size_t size)
     {
-        if (_data)
-        {
-            services::daal_free(_data);
-            _data = nullptr;
-        }
+        _size = size ? nextPowerOf2(size) : 1; //Handles both cases size == 0 and size != 0
+        _data = static_cast<T *>(services::internal::service_malloc<T, cpu>(_size * sizeof(T)));
+        _sizeMinus1 = (long) _size - 1;
+        _top = -1;
+        return _data;
     }
 
     void reset()
     {
-        _top   = _sizeMinus1;
-        _count = 0;
+        _top = -1;
     }
 
     DAAL_FORCEINLINE services::Status push(const T & value)
     {
         services::Status status;
-        if (_count >= _size)
+        if (_top >= _sizeMinus1)
         {
             status = grow();
             DAAL_CHECK_STATUS_VAR(status)
         }
-
-        _top        = (_top + 1) & _sizeMinus1;
-        _data[_top] = value;
-        ++_count;
-
+        _data[++_top] = value;
         return status;
     }
 
     DAAL_FORCEINLINE T pop()
     {
-        const T value = _data[_top--];
-        _top          = _top & _sizeMinus1;
-        --_count;
-        return value;
+        return _data[_top--];
     }
 
-    bool empty() const { return (_count == 0); }
+    bool empty() const { return _top == -1; }
 
-    size_t size() const { return _count; }
+    size_t size() const { return (size_t) (_top + 1); }
 
     services::Status grow()
     {
-        _size *= 2;
+        _size <<= 1; // Equal to multiplying by 2
         T * const newData = static_cast<T *>(services::internal::service_malloc<T, cpu>(_size * sizeof(T)));
         DAAL_CHECK_MALLOC(newData)
-        if (_count == 0)
-        {
-            _top = _size - 1;
-        }
-        _sizeMinus1 = _size - 1;
-        int result  = services::internal::daal_memcpy_s(newData, _size * sizeof(T), _data, _count * sizeof(T));
+        _sizeMinus1 = (long) _size - 1;
+        int result  = services::internal::daal_memcpy_s(newData, _size * sizeof(T), _data, size() * sizeof(T));
         T * oldData = _data;
         _data       = newData;
         services::daal_free(oldData);
         oldData = nullptr;
         return (!result) ? services::Status() : services::Status(services::ErrorMemoryCopyFailedInternal);
     }
-    const T* get_raw(void){
-        return _data;
-    }
-private:
-    T * _data;
-    size_t _top;
-    size_t _count;
-    size_t _size;
-    size_t _sizeMinus1;
 };
```

Comment:
I suggest to refuse from usage smth. like cyclic stack for several reasons:
1) This implementation has less complexity
2) Performance: I've gained small reduction of execution time
3) This implementation fixes bug with duplication of elements

Reply:
Rewrite by single line impl.

Reply:
What do you mean?